server/tests/util/test_binpacking.py

```diff
@@ -242,6 +349,67 @@ def test_balance(self, bucketlist):
             batches == new_batches
         ), "Balanced batches were not the same as original batches!"
 
+    def test_balance_skewed(self, skewedbucketlist):
+        new_bl = skewedbucketlist.balance()
+
+        assert (
+            skewedbucketlist.deviation() >= new_bl.deviation()
+        ), "Balanced list has higher deviation than original assignment! {} is greater than {}".format(
+            new_bl.deviation(), skewedbucketlist.deviation()
+        )
+
+        num_batches = sum([len(bucket.batches) for bucket in skewedbucketlist.buckets])
+        balanced_num_batches = sum([len(bucket.batches) for bucket in new_bl.buckets])
+
+        assert (
+            num_batches == balanced_num_batches
+        ), "New batch has different number of batches than expected! Got {}, expected {}".format(
+            balanced_num_batches, num_batches
+        )
+
+        batches = set()
+        for batch in [bucket.batches for bucket in skewedbucketlist.buckets]:
```

Comment:
i don't think this does quite what you intended

```python
>>> bucket1 = Bucket("1")
>>> bucket1.add_batch("a", 10)
>>> bucket1.add_batch("b", 20)
>>> bucket2 = Bucket("2")
>>> bucket2.add_batch("c", 30)
>>> bucket2.add_batch("d", 40)
>>> bbl = BalancedBucketList([bucket1, bucket2])
>>> batches = set()
>>> for batch in [bucket.batches for bucket in bbl.buckets]:
...   batches.union(set(batch.keys()))
... 
{'d', 'a'}
{'b', 'c'}
>>> batches
set()
```

does this work?

```python
>>> batches = {batch for bucket in bbl.buckets for batch in bucket.batches}
>>> batches
{'b', 'c', 'd', 'a'}
```

Reply:
fixed!